services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/BaseIcebergSchemaValidator.java

```diff
@@ -16,32 +16,63 @@
 @Component
 public class BaseIcebergSchemaValidator implements SchemaValidator {
   @Override
-  public void validateWriteSchema(
-      Schema oldSchema, Schema newSchema, Schema updatedSchema, String tableUri)
+  public void validateWriteSchema(Schema oldSchema, Schema newSchema, String tableUri)
       throws InvalidSchemaEvolutionException, IllegalArgumentException {
-    // TODO: Detect and resolve rename and reorder
-    // This implementation assumes both old and new schemas have field ids internally assigned by
-    // iceberg rather than specified by the users
+    validateFields(oldSchema, newSchema, tableUri);
     enforceDeleteColumnFailure(oldSchema, newSchema, tableUri);
-    TypeUtil.validateWriteSchema(
-        updatedSchema, oldSchema, /*checkNullability*/ true, /*checkOrdering*/ false);
   }
 
   private void enforceDeleteColumnFailure(Schema oldSchema, Schema newSchema, String tableUri) {
     // if old schema couldn't find a key in the new schema map, it should fail
     MapDifference<Integer, String> diff =
         Maps.difference(oldSchema.idToName(), newSchema.idToName());
 
-    if (diff.entriesOnlyOnLeft().size() != 0) {
+    if (!diff.entriesOnlyOnLeft().isEmpty()) {
       throw new InvalidSchemaEvolutionException(
           tableUri, newSchema.toString(), oldSchema.toString(), "Some columns are dropped");
     }
-    if (diff.entriesDiffering().size() != 0) {
+    if (!diff.entriesDiffering().isEmpty()) {
       throw new InvalidSchemaEvolutionException(
           tableUri,
           newSchema.toString(),
           oldSchema.toString(),
           "Given same id, column name is different");
     }
   }
+
+  /**
+   * Since @param newSchema is considered to be the new source-of-truth table schema, OpenHouse just
+   * need to ensure all field-Id are still matching before directly use it to assemble {@link
+   * org.apache.iceberg.TableMetadata}
+   *
+   * <p>Note that Iceberg by itself is case-sensitive by default, the casing info can be lost
+   * through the SQL engine layer. Here OH is honoring the default case sensitivity for column
+   * name-based resolution.
+   *
+   * @param oldSchema existing schema in table catalog
+   * @param newSchema user-provided schema that may contain evolution
+   * @param tableUri tableUri
+   */
+  protected void validateFields(Schema oldSchema, Schema newSchema, String tableUri) {
```

Comment:
this seems to be checking all fields from oldSchema are present in newSchema with same casing & fieldids. 
but do we also need (1) nullability matching between the fields of the two schemas? (2) do we enforce somewhere that the new top level fields added cannot be non-nullable.

Reply:
I am adding back the Iceberg validation method. That should cover the nullability case you mentioned.